packages/esm-patient-common-lib/src/workspaces/workspaces.ts

```diff
@@ -114,63 +156,51 @@ function getTitleFromExtension(ext: ExtensionRegistration) {
   const newWorkspace = {
     ...workspace,
     closeWorkspace: (ignoreChanges = true) => closeWorkspace(name, ignoreChanges),
     promptBeforeClosing: (testFcn) => promptBeforeClosing(name, testFcn),
     additionalProps,
   };
-  const existingWorkspaces = state.openWorkspaces.filter((w) => w.type == newWorkspace.type);
 
-  if (existingWorkspaces.length == 0) {
-    store.setState({ ...state, openWorkspaces: [newWorkspace, ...state.openWorkspaces] });
-  } else {
-    const existingIdx = state.openWorkspaces.findIndex((w) => w.name == name);
-    const promptCheckFcn = getPromptBeforeClosingFcn(existingWorkspaces[0].name);
-    if (existingIdx >= 0) {
-      const restOfWorkspaces = [...state.openWorkspaces];
-      restOfWorkspaces.splice(existingIdx, 1);
-      state.openWorkspaces[existingIdx].additionalProps = newWorkspace.additionalProps;
-      const openWorkspaces = [state.openWorkspaces[existingIdx], ...restOfWorkspaces];
-      store.setState({ ...state, openWorkspaces });
-    } else if (!promptCheckFcn || promptCheckFcn()) {
-      const currentName = existingWorkspaces[0].title ?? existingWorkspaces[0].name;
-      const prompt: Prompt = {
-        title: translateFrom(
-          '@openmrs/esm-patient-chart-app',
-          'activeFormWarning',
-          'There is an active form open in the workspace',
-        ),
-        body: translateFrom(
-          '@openmrs/esm-patient-chart-app',
-          'workspaceModalText',
-          `Launching a new form in the workspace could cause you to lose unsaved work on the ${currentName} form.`,
-          { formName: currentName },
-        ),
-        onConfirm: () => {
-          const state = store.getState();
-          store.setState({
-            openWorkspaces: [newWorkspace, ...state.openWorkspaces.filter((w) => w.type != newWorkspace.type)],
-            prompt: null,
-          });
-        },
-        confirmText: translateFrom('@openmrs/esm-patient-chart-app', 'openAnyway', 'Open anyway'),
-      };
-      store.setState({ ...state, prompt });
-    } else {
-      const state = store.getState();
-      store.setState({
-        ...state,
-        openWorkspaces: [newWorkspace, ...state.openWorkspaces.filter((w) => w.type != newWorkspace.type)],
-      });
+  const updateStoreWithNewWorkspace = (workspaceToBeAdded: OpenWorkspace, restWorkspaces = null) => {
+    if (!restWorkspaces) {
+      restWorkspaces = store.getState().openWorkspaces;
     }
-  }
-  if (store.getState().workspaceWindowState === 'hidden') {
-    updateWorkspaceWindowState(
-      store.getState().openWorkspaces[0].preferredWindowSize === 'maximized' ? 'maximized' : 'normal',
-    );
+    store.setState((state) => ({ ...state, openWorkspaces: [workspaceToBeAdded, ...restWorkspaces] }));
+    if (store.getState().workspaceWindowState === 'hidden') {
+      updateWorkspaceWindowState(
+        store.getState().openWorkspaces[0].preferredWindowSize === 'maximized' ? 'maximized' : 'normal',
+      );
+    }
```

Comment:
Can we work this update logic into the `setState()` cal above? The idea here is to minimise the number of calls to `getState()` to minimise the chances of getting inconsistent results, i.e.,

```js
store.setState((state) => {
  let workspaceWindowState = state.workspaceWindowState;
  if (workspaceWindowState === 'hidden') {
    workspaceWindowState = workspaceToBeAdded.preferredWindowSize === 'maximized' ? 'maximized' : 'normal';
  }
  return { ...state, openWorkspaces: [workspaceToBeADded, ...(restWorkspaces ?? state.openWorkspaces)], workspaceWindowState };
});
```

(Notice I've also moved the `restWorkspaces` into here. Effectively, one call to `getState()` and then a single update to the store state).

There are a couple of reasons for this:

1. As above, ensuring we have a "consistent" (really, "transactional") view of the workspace state.
2. We only update the state once, as a single entity. Every call to `setState()` will ultimately result in an update to the state and then that update being sent to all state subscribers; in essence, this can help avoid multiple React re-renders for a single state change operation.

Reply:
Right, I totally missed it out.
Thanks for catching this!!!

Reply:
Please re-review.
Thanks!